Add ResourceQuota doc (#346)
harvester/docs: docs/rancher/resource-quota.md

For more information, see [Memory Overhead](https://kubevirt.io/user-guide/virtual_machines/virtual_hardware/#memory-overhead).

For more information on how the memory overhead is calculated in Kubevirt, refer to [kubevirt/pkg/virt-controller/services/template.go](https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/services/template.go#L1804).

In order to make it much clearer how Memory Overhead is calculated, it is described by a simple example.
We assume that a new VM will be created with the following configuration:

- CPU: 8 cores
- Memory: 16Gi

Memory Overhead is calculated in 4 sections:

- Memory PageTables overhead (one bit for every 512b of RAM size)，i.e. 16Gi memory requires 32Mi overhead.

```go
pagetableMemory := resource.NewScaledQuantity(vmiMemoryReq.ScaledValue(resource.Kilo), resource.Kilo)
pagetableMemory.Set(pagetableMemory.Value() / 512)
overhead.Add(*pagetableMemory)
```

> - VM fixed overhead 180Mi (branch master: 210Mi)：
> 
> ```go
> VirtLauncherMonitorOverhead = "25Mi" // The `ps` RSS for virt-launcher-monitor
> VirtLauncherOverhead        = "75Mi" // The `ps` RSS for the virt-launcher process
> VirtlogdOverhead            = "17Mi" // The `ps` RSS for virtlogd
> LibvirtdOverhead            = "33Mi" // The `ps` RSS for libvirtd
> QemuOverhead                = "30Mi" // The `ps` RSS for qemu, minus the RAM of its (stressed) guest, minus the virtual page table
> 
> overhead.Add(resource.MustParse(VirtLauncherMonitorOverhead))
> overhead.Add(resource.MustParse(VirtLauncherOverhead))
> overhead.Add(resource.MustParse(VirtlogdOverhead))
> overhead.Add(resource.MustParse(LibvirtdOverhead))
> overhead.Add(resource.MustParse(QemuOverhead))
> ```

Reviewer: ```suggestion
- VM Fixed Overhead: This consists of several components:
    - `VirtLauncherMonitorOverhead`: 25Mi  (the `ps` RSS for virt-launcher-monitor)
    - `VirtLauncherOverhead`: 75Mi  (the `ps` RSS for the virt-launcher process)
    - `VirtlogdOverhead`: 17Mi  (the `ps` RSS for virtlogd)
    - `LibvirtdOverhead`: 33Mi. (the `ps` RSS for libvirtd)
    - `QemuOverhead` : 30Mi (the `ps` RSS for qemu, minus the RAM of its (stressed) guest, minus the virtual page table)
```

> - 8Mi per CPU(vCPU) overhead and in addition to a fixed 8Mi IOTread overhead.
> 
> ```go
> // vCPU calculated
> coresMemory := resource.MustParse("8Mi")
> value := coresMemory.Value() * vcpus
> coresMemory = *resource.NewQuantity(value, coresMemory.Format)
> overhead.Add(coresMemory)
> // static overhead for IOThread
> overhead.Add(resource.MustParse("8Mi"))
> ```

Reviewer: ```suggestion
- 8Mi per CPU (vCPU) Overhead: Additionally, 8Mi of overhead per vCPU is added, along with a fixed 8Mi overhead for IOTread:
```

> - The extra added overhead，e.g. : video RAM overhead, arch overhead etc. see [Additional Overhead](https://github.com/kubevirt/kubevirt/blob/2bb88c3d35d33177ea16c0f1e9fffdef1fd350c6/pkg/virt-controller/services/template.go#L1853-L1890)

Reviewer: ```suggestion
- Extra Added Overhead: This encompasses various factors like video RAM overhead and architecture overhead. Refer to [Additional Overhead](https://github.com/kubevirt/kubevirt/blob/2bb88c3d35d33177ea16c0f1e9fffdef1fd350c6/pkg/virt-controller/services/template.go#L1853-L1890) for further details.
```

> The calculation shows that the VM instance needs to add the memory overhead of about 276Mi.

Reviewer: ```suggestion
This calculation demonstrates that the VM instance necessitates an additional memory overhead of approximately 276Mi.
```

## Adjust ResourceQuota automatically during the migration
When the resource quota defined by the `ResourceQuota` object is exhausted, you cannot migrate the virtual machine. During the migration, a pod with the same resources as the source VM will be automatically created. If the resources required for creating the pod exceed the quota configured, the migration cannot proceed.

In Harvester, before the migration, the values of `ResourceQuota` are automatically increased based on the resources required by the target virtual machine. After the migration, `ResourceQuotas` are restored (reduced) to their previous values.

Keep the considerations below in mind about automatically adjusting `ResourceQuota`:

- When the value of `ResourceQuota` is being increased, if you create, start, or restore other VMs, Harvester will verify if the resources are sufficient based on the original `ResourceQuota`. If not, the system will prompt that the migration cannot be completed.
- `ResourceQuota` cannot be changed during VM migration.

> - After `ResourceQuota` is increased, there may be resource competition between non-VM pods and VM pods, resulting in migration failure. Therefore, we don't recommend deploying hybrid pods in the same namespace.

Reviewer: ```suggestion
- After expanding `ResourceQuota`, potential resource contention may emerge between non-VM pods and VM pods, leading to migration failures. Consequently, deploying non-VM pods to the same namespace is not recommended.
```

> - Due to the webhook validator's unreliability, the VM Controller performs a second validation to verify resource sufficiency. If the resources are insufficient, the `RunStrategy` of the VM will become **Halted**, and the annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM custom resource, informing you that the VM was shut down due to insufficient resources.

Reviewer: ```suggestion
- Due to the concurrent limitation of the webhook validator, the VM controller will execute a secondary validation to confirm resource sufficiency. If the resources are insufficient, it will auto config VM's`RunStrategy` to **Halted**, and a new annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM object, informing you that the VM was shut down due to insufficient resources.
```

---
sidebar_position: 6
sidebar_label: Resource Quota
title: "Resource Quota"
keywords:
  - Harvester
  - harvester
  - Rancher
  - rancher
  - Resource Quota
Description: ResourceQuota allows administrators to set resource limits per namespace, preventing excessive resource usage and ensuring the smooth operation of other namespaces when the quota is reached.
---

> [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/) is a resource quota management mechanism in Kubernetes used to limit the usage of resources within a namespace. It helps administrators control and restrict the allocation of cluster resources to ensure fairness and controlled resource distribution among namespaces.

Reviewer 2: ```suggestion
[ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/) is used to limit the usage of resources within a namespace. It helps administrators control and restrict the allocation of cluster resources to ensure fairness and controlled resource distribution among namespaces.
```

> ResourceQuota can define quota limits for the following resources:

Reviewer 2: ```suggestion
ResourceQuota can define usage limits for the following resources:
```

- CPU (compute resources): Limits the usage of compute resources in terms of CPU cores or CPU time.
- Memory (memory resources): Limits the usage of memory resources in bytes or other recognizable memory units.
Using `ResourceQuota`, administrators can set different resource limits for each namespace, ensuring the proper utilization and allocation of resources. When a namespace reaches its resource quota limit, further resource requests are denied, preventing excessive consumption of cluster resources and ensuring the normal operation of other namespaces.

In summary, `ResourceQuota` is a mechanism in Kubernetes used to restrict and manage the usage of resources within namespaces, promoting fair resource sharing and proper allocation.

## ResourceQuota in Harvester
In multi-cluster management mode, Harvester supports administrators in setting resource quotas for namespaces, as follows:

- Click `⋮` -> `Virtualization Management` -> `Import Harvester Cluster` -> `Project/Namespace`.
- Click `Create Project`.

> - Enter the project name, click `Resource Quotas` and Add Resource Type, select `CPU Limit` or `Memory Limit` and enter the quota value, and click `Create`.

Reviewer 2: The steps are modified according to the actual steps found in the Rancher UI. Please help double-check if I understand it correctly ^^

  ![](/img/v1.2/rancher/create-project.png)
- Under the newly created Project, click `Create Namespace`.

> - Enter the Namespace name, which supports adjusting for the default quota, and click `Create`.

Reviewer 2: I guess these two steps are about modifying the quotas? So I separate them into another independent task

  ![](/img/v1.2/rancher/create-namespace.png)

## Overhead Memory of Virtual Machine
When creating a virtual machine, the virtual machine controller will automatically add overhead resources to the virtual machine. The overhead resources are used to ensure the regular operation of the virtual machine, and the configuration of memory limits requires consideration of reserving more memory.

For more information, see the [Memory Overhead documentation](https://kubevirt.io/user-guide/virtual_machines/virtual_hardware/#memory-overhead).

For more information on how to calculate memory overhead, refer to [Code](https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/services/template.go#L1804).

## Adjust ResourceQuota automatically during migration

> When the resource quota defined by the `ResourceQuota` object is exhausted, virtual machine migration is not supported. Migration requires creating a pod with the same resources as the migrated VM. If the resources required for creating the pod exceed the quota, migration cannot proceed.

Reviewer 2: ```suggestion
When the resource quota defined by the `ResourceQuota` object is exhausted, you cannot migrate the virtual machine. To perform the migration, you must create a pod with the same resources as the VM you are about to migrate. If the resources required for creating the pod exceed the quota configured, the migration cannot proceed.
```

> In Harvester, before migration, the resource limits of `ResourceQuota` are increased based on the specifications of the target virtual machine. After the migration, the resource limits are restored (reduced) to their previous values.

Reviewer 2: ```suggestion
In Harvester, before the migration, the values of `ResourceQuota` are automatically increased based on the resources required by the target virtual machine. After the migration, `ResourceQuotas` are restored (reduced) to their previous values.
```

Reviewer 2: `resource limit` is actually the same as `ResourceQuota`? if so, prefer using a consistent term

> There are certain considerations when automatically adjusting `ResourceQuota`:

Reviewer 2: ```suggestion
Keep the considerations below in mind about automatically adjusting `ResourceQuota`:
```

> - During the resource increase period of `ResourceQuota`, if users are creating, starting, or restoring VMs, Harvester will verify if the original resources are sufficient. If there are insufficient resources, the system will prompt the user that the operation cannot be completed.

Reviewer 2: ```suggestion
- When the value of `ResourceQuota` is being increased, if you create, start, or restore other VMs, Harvester will verify if the resources are sufficient based on the original `ResourceQuota`. If not, the system will prompt that the migration cannot be completed.
```

- `ResourceQuota` cannot be changed during VM migration.

> - After increasing `ResourceQuota`, there may be resource competition between non-VM pods and VM pods, resulting in migration failure. Therefore, deploying Hybrid Pods in the same namespace is not recommended.

Reviewer 2: ```suggestion
- After `ResourceQuota` is increased, there may be resource competition between non-VM pods and VM pods, resulting in migration failure. Therefore, we don't recommend deploying hybrid pods in the same namespace.
```

> - Due to the webhook validator's unreliability, the VM Controller performs a second validation to verify if the resources are sufficient. If there is insufficient resources, the RunStrategy of the VM will change to **Halted**, and the annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` has been added to the VirtualMachine custom resource to let you know the VM was shutdown due to insufficient resources.

Reviewer 2: ```suggestion
- Due to the webhook validator's unreliability, the VM Controller performs a second validation to verify resource sufficiency. If the resources are insufficient, the `RunStrategy` of the VM will become **Halted**, and the annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM custom resource, informing you that the VM was shut down due to insufficient resources.
```

---
sidebar_position: 6
sidebar_label: Resource Quota
title: "Resource Quota"
keywords:
  - Harvester
  - harvester
  - Rancher
  - rancher
  - Resource Quota
Description: ResourceQuota allows administrators to set resource limits per namespace, preventing excessive resource usage and ensuring the smooth operation of other namespaces when the quota is reached.
---

[ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/) is used to limit the usage of resources within a namespace. It helps administrators control and restrict the allocation of cluster resources to ensure fairness and controlled resource distribution among namespaces.

> In Harvester, ResourceQuota can define usage limits for the following resources:
> 
> - CPU: Limits the usage of compute resources, including CPU cores and CPU time.
> - Memory: Limits the usage of memory resources in bytes or other recognizable memory units.
> Using `ResourceQuota`, administrators can set different resource limits for each namespace, ensuring the proper utilization and allocation of resources. When a namespace reaches its resource quota limit, further resource requests are denied, preventing excessive consumption of cluster resources and ensuring the normal operation of other namespaces.

Reviewer 3: ```suggestion
In Harvester, ResourceQuota can define usage limits for the following resources:
- CPU: Limits compute resource usage, including CPU cores and CPU time.
- Memory: Limits the usage of memory resources in bytes or other recognizable memory units.
Using `ResourceQuota`, administrators can set different resource limits for each namespace, ensuring the proper utilization and allocation of resources. When a namespace reaches its resource quota limit, further resource requests are denied, preventing excessive consumption of cluster resources and ensuring the regular operation of other namespaces.
```

## Set ResourceQuota in Rancher
In the multi-cluster management UI, administrators can configure resource quotas for namespaces through the following steps:

> 1. In the Rancher UI, click on the hamburger menu, and then click **Virtualization Management** > **Import Cluster** to import a Harvester cluster (see [Import Cluster](https://docs.harvesterhci.io/v1.1/rancher/virtualization-management)).

Reviewer 3: ```suggestion
1. In the Rancher UI, go to the hamburger menu, and then select **Virtualization Management** > **Import Cluster** to import a Harvester cluster. See [Import Cluster](https://docs.harvesterhci.io/v1.1/rancher/virtualization-management).
```

For more information, see [Memory Overhead](https://kubevirt.io/user-guide/virtual_machines/virtual_hardware/#memory-overhead).

For more information on how the memory overhead is calculated in Kubevirt, check the code [here](https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/services/template.go#L1804).

## Adjust ResourceQuota automatically during the migration
When the allocated resource quota governed by the `ResourceQuota` object reaches its limit, migrating a virtual machine becomes unfeasible. The migration process involves the automatic creation of a new pod mirroring the resource requirements of the source virtual machine. If these pod creation prerequisites surpass the defined quota, the migration operation cannot proceed.

```suggestion
_Available as of v1.2.0_

> In Harvester, the `ResourceQuota` values are dynamically expanded ahead of migration to accommodate the resource needs of the target virtual machine. Subsequent to migration, the ResourceQuotas are reinstated to their prior configurations.
> 
> Please be aware of the following considerations of the automatic resizing of `ResourceQuota`:
> 
> - During the process of raising the `ResourceQuota` value, if you create, start, or restore other VMs, Harvester will verify if the resources are sufficient based on the original `ResourceQuota`. If the conditions are not met, the system will alerts that the migration process is not feasible.
> - `ResourceQuota` cannot be changed during VM migration.
> - After expanding `ResourceQuota`, potential resource contention may emerge between non-VM pods and VM pods, leading to migration failures. Consequently, deploying non-VM pods to the same namespace is not recommended.
> - Due to the concurrent limitation of the webhook validator, the VM controller will execute a secondary validation to confirm resource sufficiency. If the resources are insufficient, it will auto config VM's`RunStrategy` to **Halted**, and a new annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM object, informing you that the VM was shut down due to insufficient resources.
>   ![](/img/v1.2/rancher/vm-annotation-insufficient-resource-quota.png)

Reviewer 3: ```suggestion
In Harvester, the `ResourceQuota` values dynamically expand ahead of migration to accommodate the resource needs of the target virtual machine. After migration, the ResourceQuotas are reinstated to their prior configurations.

Please be aware of the following considerations of the automatic resizing of `ResourceQuota`:
- When raising the `ResourceQuota` value, if you create, start, or restore other VMs, Harvester will verify if the resources are sufficient based on the original `ResourceQuota`. If the conditions are not met, the system will alert that the migration process is not feasible.
- `ResourceQuota` cannot be changed during VM migration.
- After expanding `ResourceQuota`, potential resource contention may emerge between non-VM pods and VM pods, leading to migration failures. Consequently, deploying non-VM pods to the same namespace is not recommended.
- Due to the concurrent limitation of the webhook validator, the VM controller will execute a secondary validation to confirm resource sufficiency. If the resources are insufficient, it will auto config VM's`RunStrategy` to **Halted**, and a new annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM object, informing you that the VM was shut down due to insufficient resources.
  ![](/img/v1.2/rancher/vm-annotation-insufficient-resource-quota.png)
```

---
sidebar_position: 6

> sidebar_label: Resource Quota

Reviewer 3: ```suggestion
sidebar_label: Resource Quotas
```

> title: "Resource Quota"

Reviewer 3: ```suggestion
title: "Resource Quotas"
```

keywords:
  - Harvester
  - harvester
  - Rancher
  - rancher
  - Resource Quota
Description: ResourceQuota allows administrators to set resource limits per namespace, preventing excessive resource usage and ensuring the smooth operation of other namespaces when the quota is reached.
---

[ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/) is used to limit the usage of resources within a namespace. It helps administrators control and restrict the allocation of cluster resources to ensure fairness and controlled resource distribution among namespaces.

In Harvester, ResourceQuota can define usage limits for the following resources:

> - CPU: Limits compute resource usage, including CPU cores and CPU time.

Reviewer 3: ```suggestion
- **CPU:** Limits compute resource usage, including CPU cores and CPU time.
```

> - Memory: Limits the usage of memory resources in bytes or other recognizable memory units.

Reviewer 3: ```suggestion
- **Memory:** Limits the usage of memory resources in bytes or other recognizable memory units.
```

1. Specify the desired project **Name**. Next, go to the **Resource Quotas** tab and select the **Add Resource** option. Within the **Resource Type** field, select either **CPU Limit** or **Memory Limit** and define the **Project Limit** and **Namespace Default Limit** values.
  ![](/img/v1.2/rancher/create-project.png)

You can configure the **Namespace** limits as follows: 

1. Find the newly created project, and select **Create Namespace**.
1. Specify the desired namespace **Name**, and adjust the limits.
1. Complete the process by selecting **Create**.
  ![](/img/v1.2/rancher/create-namespace.png)

## Overhead Memory of Virtual Machine
Upon creating a virtual machine (VM), the VM controller seamlessly incorporates overhead resources into the VM's configuration. These additional resources intend to guarantee the consistent and uninterrupted functioning of the VM. It's important to note that configuring memory limits requires a higher memory reservation due to the inclusion of these overhead resources.

For example, consider the creation of a new VM with the following configuration:
- CPU: 8 cores
- Memory: 16Gi

:::note

> The operating system does not affect overhead calculations, either Linux or Windows.

Reviewer 3: ```suggestion
The operating system, either Linux or Windows, does not affect overhead calculations.
```

:::

Memory Overhead is calculated in the following sections:

> - Memory PageTables Overhead: This accounts for one bit for every 512b RAM size. For instance, a memory of 16Gi requires an overhead of 32Mi.

Reviewer 3: ```suggestion
- **Memory PageTables Overhead:** This accounts for one bit for every 512b RAM size. For instance, a memory of 16Gi requires an overhead of 32Mi.
```

> - VM Fixed Overhead: This consists of several components:

Reviewer 3: ```suggestion
- **VM Fixed Overhead:** This consists of several components:
```

    - `VirtLauncherMonitorOverhead`: 25Mi  (the `ps` RSS for virt-launcher-monitor)
    - `VirtLauncherOverhead`: 75Mi  (the `ps` RSS for the virt-launcher process)
    - `VirtlogdOverhead`: 17Mi  (the `ps` RSS for virtlogd)

>     - `LibvirtdOverhead`: 33Mi. (the `ps` RSS for libvirtd)

Reviewer 3: ```suggestion
    - `LibvirtdOverhead`: 33Mi (the `ps` RSS for libvirtd)
```

    - `QemuOverhead` : 30Mi (the `ps` RSS for qemu, minus the RAM of its (stressed) guest, minus the virtual page table)

> - 8Mi per CPU (vCPU) Overhead: Additionally, 8Mi of overhead per vCPU is added, along with a fixed 8Mi overhead for IOThread.

Reviewer 3: ```suggestion
- **8Mi per CPU (vCPU) Overhead:** Additionally, 8Mi of overhead per vCPU is added, along with a fixed 8Mi overhead for IOThread.
```

> - Extra Added Overhead: This encompasses various factors like video RAM overhead and architecture overhead. Refer to [Additional Overhead](https://github.com/kubevirt/kubevirt/blob/2bb88c3d35d33177ea16c0f1e9fffdef1fd350c6/pkg/virt-controller/services/template.go#L1853-L1890) for further details.

Reviewer 3: ```suggestion
- **Extra Added Overhead:** This encompasses various factors like video RAM overhead and architecture overhead. Refer to [Additional Overhead](https://github.com/kubevirt/kubevirt/blob/2bb88c3d35d33177ea16c0f1e9fffdef1fd350c6/pkg/virt-controller/services/template.go#L1853-L1890) for further details.
```

_Available as of v1.2.0_

In Harvester, the `ResourceQuota` values dynamically expand ahead of migration to accommodate the resource needs of the target virtual machine. After migration, the ResourceQuotas are reinstated to their prior configurations.

Please be aware of the following considerations of the automatic resizing of `ResourceQuota`:
- When raising the `ResourceQuota` value, if you create, start, or restore other VMs, Harvester will verify if the resources are sufficient based on the original `ResourceQuota`. If the conditions are not met, the system will alert that the migration process is not feasible.
- `ResourceQuota` cannot be changed during VM migration.
- After expanding `ResourceQuota`, potential resource contention may emerge between non-VM pods and VM pods, leading to migration failures. Consequently, deploying non-VM pods to the same namespace is not recommended.

> - Due to the concurrent limitation of the webhook validator, the VM controller will execute a secondary validation to confirm resource sufficiency. If the resources are insufficient, it will auto config VM's`RunStrategy` to **Halted**, and a new annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM object, informing you that the VM was shut down due to insufficient resources.

Reviewer 3: ```suggestion
- Due to the concurrent limitation of the webhook validator, the VM controller will execute a secondary validation to confirm resource sufficiency. If the resources are insufficient, it will auto config the VM's`RunStrategy` to **Halted**, and a new annotation `harvesterhci.io/insufficient-resource-quota: xxxxx` will be added to the VM object, informing you that the VM was shut down due to insufficient resources.
```
